### ISSUE TITLE
Unable to display data returned from Jenkins

### ISSUE THREAD

**Issue by kalpik** [marcusolsson/grafana-jsonapi-datasource#15]
Hello!

I'm trying to parse this data:

`{
  "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowJob",
  "builds" : [
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 220,
      "result" : "SUCCESS"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 219,
      "result" : "UNSTABLE"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 218,
      "result" : "ABORTED"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 217,
      "result" : "UNSTABLE"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 216,
      "result" : "UNSTABLE"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 215,
      "result" : "ABORTED"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 214,
      "result" : "UNSTABLE"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 213,
      "result" : "ABORTED"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 212,
      "result" : "ABORTED"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 211,
      "result" : "UNSTABLE"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 210,
      "result" : "ABORTED"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 209,
      "result" : "SUCCESS"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 208,
      "result" : "UNSTABLE"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 207,
      "result" : "ABORTED"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 206,
      "result" : "ABORTED"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 205,
      "result" : "SUCCESS"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 204,
      "result" : "SUCCESS"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 203,
      "result" : "SUCCESS"
    },
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 202,
      "result" : "ABORTED"
    }
  ]
}`

Trying to use these queries, but I get an error. I'm sure it's something stupid!
![image](https://user-images.githubusercontent.com/481869/99689644-85e10400-2a87-11eb-9ac8-86d18c7703ce.png)

**Comment by kalpik** [comment]
If I only have $.builds[*].number, it shows the build numbers, but somehow I cannot display result.

**Comment by marcusolsson** [comment]
The JSON you posted seems to work for me:

<img width="871" alt="Screen Shot 2020-11-19 at 21 33 21" src="https://user-images.githubusercontent.com/8396880/99721023-db7cd700-2aae-11eb-956c-5987dc8662df.png">

The error you get tells me that the result of each query don't contain the same number of rows. I'm guessing that one of the `number` and `result` properties is sometimes missing from the builds.

**Comment by kalpik** [comment]
Thanks for your reply. That's what I thought as well, but I just cannot get any result to show! Even if I do $.builds[0].result, I get nothing. I can get $.builds.[0].number. So really not sure what's happening here. I checked the response from the Jenkins API, and it has both number and result for all elements, the JSON I posted above is directly from the Jenkins API. So kind of strange it works for you, but not for me!

**Comment by marcusolsson** [comment]
I'm not sure either why it's failing for you. Could you perhaps send me the result from the Query inspector. That way I can be sure we're looking at the same data.

1. Click the Query inspector button in the upper right corner of the query editor
1. Click **Refresh** to make a new request
1. Click **Copy to clipboard**
1. Paste it in this issue

**Make sure that the request doesn't contain any sensitive information before pasting it**

**Comment by kalpik** [comment]
Thanks. But that's the other thing. In query inspector, when I click refresh, it keeps loading, which I again found to be strange.

Keeps being stuck on this:

![image](https://user-images.githubusercontent.com/481869/99784331-f33e7480-2b1b-11eb-8e7f-51bfa25d4bbf.png)

**Comment by kalpik** [comment]
Okay, figured it out. I need to pass the URL query parameters as URL encoded, else it doesn't work. Maybe something you can fix? :)

**Comment by marcusolsson** [comment]
Oh! That's indeed something I can fix! 😅

**Comment by marcusolsson** [comment]
Can I ask you to provide the query parameters that broke things for you, what you had to put in there instead?

**Comment by kalpik** [comment]
Here you go:

URL: https://my-jenkins-instance.com/job/Cloud/job/attempt_iqa_promotion/job/develop/api/json

Query Parameters actual (which broke): pretty=true&tree=builds[number,result,url]{0,10}
URL Encoded (which work): pretty=true&tree=builds%5Bnumber%2Cresult%2Curl%5D%7B0%2C10%7D

Note that if I also encode '=' and '&', it doesn't work again. So only some parts need to be URL encoded.

**Comment by kubala156** [comment]
Hey i have the same situation and my URL dose not need to be encoded (http://api.gios.gov.pl/pjp-api/rest/station/findAll)
Sample JSON:

```
[
    {
        "id": 114,
        "stationName": "Wrocław - Bartnicza",
        "gegrLat": "51.115933",
        "gegrLon": "17.141125",
        "city": {
            "id": 1064,
            "name": "Wrocław",
            "commune": {
                "communeName": "Wrocław",
                "districtName": "Wrocław",
                "provinceName": "DOLNOŚLĄSKIE"
            }
        },
        "addressStreet": "ul. Bartnicza"
    },
    {
        "id": 117,
        "stationName": "Wrocław - Korzeniowskiego",
        "gegrLat": "51.129378",
        "gegrLon": "17.029250",
        "city": {
            "id": 1064,
            "name": "Wrocław",
            "commune": {
                "communeName": "Wrocław",
                "districtName": "Wrocław",
                "provinceName": "DOLNOŚLĄSKIE"
            }
        },
        "addressStreet": "ul. Wyb. J.Conrada-Korzeniowskiego 18"
    }
]
```

**Comment by marcusolsson** [comment]
@kubala156 I can reproduce it the JSON you've sent me. This is the result from the data source. There are two reasons why this doesn't work:

1. `$.[*].id` matches recursively (just learned this myself), which also adds the city ID to the id values. I fixed it by changing it to `$.[:].id`. We should document this in the README, but I need to read up on it a little more.
1. `gegrLong` gets parsed as a timestamp for some reason. This is a bug :)

```json
[
  {
    "name": "id",
    "type": "number",
    "values": [
      114,
      117,
      1064,
      1064
    ]
  },
  {
    "name": "stationName",
    "type": "string",
    "values": [
      "Wrocław - Bartnicza",
      "Wrocław - Korzeniowskiego"
    ]
  },
  {
    "name": "gegrLat",
    "type": "string",
    "values": [
      "51.115933",
      "51.129378"
    ]
  },
  {
    "name": "gegrLong",
    "type": "time",
    "values": []
  }
]
```

**Comment by marcusolsson** [comment]
Closing this with a fix for @kalpik's issue where query parameters aren't encoded properly.

**Comment by marcusolsson** [comment]
Fixed in [v0.6.2](https://grafana.com/grafana/plugins/marcusolsson-json-datasource?version=0.6.2).

**Comment by kalpik** [comment]
Thank you :)
